linera-indexer/src/lib.rs

```diff
@@ -0,0 +1,146 @@
+{
+    /// Loads all indexer plugins
+    pub async fn load(
+        plugins: &[&str],
+        client: DB,
+        start: BlockHeight,
+        node: String,
+    ) -> Result<Self, IndexerError> {
+        let operations = if plugins.contains(&"operations") {
+            let context = ContextFromDb::create(client.clone(), vec![1], ())
+                .await
+                .map_err(|e| IndexerError::ViewError(e.into()))?;
+            Some(OperationsPlugin::load(context).await?)
+        } else {
+            None
+        };
+        let context = ContextFromDb::create(client, vec![0], ())
+            .await
+            .map_err(|e| IndexerError::ViewError(e.into()))?;
+        let state = Arc::new(Mutex::new(StateView::load(context).await?));
+        Ok(Indexer {
+            operations,
+            state,
+            start,
+            node,
+        })
+    }
+
+    /// Processes one hashed value
+    pub async fn process_value(&self, value: &HashedValue) -> Result<(), IndexerError> {
+        if let Some(plugin) = &self.operations {
+            plugin.register(value).await?
+        };
+        let chain_id = value.inner().chain_id();
+        info!("update state of chain {:?}: {:?}", chain_id, value.hash());
+        let mut state = self.state.lock().await;
+        state.chains.insert(&chain_id, value.hash())?;
+        state.save().await.map_err(IndexerError::ViewError)
+    }
+
+    /// Gets one hashed value from the node service
+    pub async fn get_value(
+        &self,
+        chain_id: ChainId,
+        hash: CryptoHash,
+    ) -> Result<HashedValue, IndexerError> {
+        let client = reqwest::Client::new();
+        let variables = block::Variables {
+            hash: Some(hash),
+            chain_id,
+        };
+        let response = post_graphql::<Block, _>(&client, &self.node, variables).await?;
+        match response.data {
+            None => Err(IndexerError::NullData(response.errors)),
+            Some(data) => match data.block {
+                Some(block) => Ok(block.into()),
+                None => Err(IndexerError::NotFound(hash)),
+            },
+        }
+    }
+
+    /// Processes one hashed value doing recursion to until the last state registered is reached
+    #[async_recursion]
+    pub async fn process(&self, value: &HashedValue) -> Result<(), IndexerError> {
+        let chain_id = value.inner().chain_id();
+        let last_block_registered = self.state.lock().await.chains.get(&chain_id).await?;
+        match value.inner().height().cmp(&self.start) {
+            Ordering::Equal => self.process_value(value).await,
+            Ordering::Greater => match value.inner().executed_block().block.previous_block_hash {
+                None => Ok(()),
+                Some(hash) => {
+                    if Some(hash) != last_block_registered {
+                        let previous_value = self.get_value(chain_id, hash).await?;
+                        self.process(&previous_value).await?
```

Comment:
Won't this overflow the stack for long enough chains?

Reply:
it would, I can see several solutions:
- only keeping the hash in the recursion,  but it would require 2 calls to the node 
- adding a indexing height -> hash for blocks allowing to crawl the chain forward, or adding a `next_block` field on top of certificateValue

Reply:
Or just make it a loop instead of recursion?

Reply:
...indeed, sorry I am to used to functional programming